File: lib/todoist.js
// Todoist /api/v1 client with cursor pagination.
// Note: /rest/v2 is deprecated; new endpoints return {results, next_cursor}.

const API_BASE = "https://api.todoist.com/api/v1";

export class TodoistError extends Error {
  constructor(message, status) {
    super(message);
    this.status = status;
  }
}

function authHeaders(token) {
  return { Authorization: `Bearer ${token}` };
}

async function paginate(url, token, params = {}) {
  const out = [];
  let cursor = null;
  for (let i = 0; i < 50; i++) {
    const u = new URL(url);
    for (const [k, v] of Object.entries(params)) u.searchParams.set(k, v);
    if (cursor) u.searchParams.set("cursor", cursor);
    const r = await fetch(u, { headers: authHeaders(token) });
    if (!r.ok) {
      throw new TodoistError(`Todoist ${r.status} ${r.statusText}`, r.status);
    }
    const body = await r.json();
    if (Array.isArray(body.results)) out.push(...body.results);
    cursor = body.next_cursor || null;
    if (!cursor) return out;
  }
  return out;
}

export async function findProjectId(token, name) {
  const projects = await paginate(`${API_BASE}/projects`, token);
  const target = name.trim().toLowerCase();
  const hit = projects.find((p) => (p.name || "").toLowerCase() === target);
  if (!hit) throw new TodoistError(`Todoist project '${name}' not found`);
  return hit.id;
}

export async function getActiveItems(token, projectId) {
  const tasks = await paginate(`${API_BASE}/tasks`, token, { project_id: projectId });
  return tasks.map((t) => ({ id: t.id, content: t.content }));
}

export async function closeTask(token, taskId) {
  const r = await fetch(`${API_BASE}/tasks/${taskId}/close`, {
    method: "POST",
    headers: authHeaders(token),
  });
  if (!r.ok) throw new TodoistError(`close ${taskId}: ${r.status}`, r.status);
}

File: src/todoist.py
import httpx

from .config import settings

# Todoist deprecated /rest/v2 in favor of unified /api/v1 (paginated responses).
API_BASE = "https://api.todoist.com/api/v1"


class TodoistError(RuntimeError):
    pass


def _headers() -> dict[str, str]:
    if not settings.todoist_api_token:
        raise TodoistError("TODOIST_API_TOKEN is not set")
    return {"Authorization": f"Bearer {settings.todoist_api_token}"}


async def _paginate(client: httpx.AsyncClient, url: str, params: dict | None = None) -> list[dict]:
    """Walk Todoist's `next_cursor` pagination until exhausted."""
    out: list[dict] = []
    cursor: str | None = None
    base_params = dict(params or {})
    while True:
        p = dict(base_params)
        if cursor:
            p["cursor"] = cursor
        r = await client.get(url, headers=_headers(), params=p)
        r.raise_for_status()
        body = r.json()
        out.extend(body.get("results", []))
        cursor = body.get("next_cursor")
        if not cursor:
            break
    return out


async def get_project_id(client: httpx.AsyncClient, name: str) -> str:
    projects = await _paginate(client, f"{API_BASE}/projects")
    for p in projects:
        if p["name"].lower() == name.lower():
            return p["id"]
    raise TodoistError(f"Todoist project '{name}' not found")


async def get_grocery_items() -> list[dict]:
    """Return active (non-completed) tasks from the configured grocery project.

    Each item: {"id": str, "content": str}
    """
    async with httpx.AsyncClient(timeout=15.0) as client:
        project_id = await get_project_id(client, settings.todoist_project_name)
        tasks = await _paginate(
            client, f"{API_BASE}/tasks", params={"project_id": project_id}
        )
        return [{"id": t["id"], "content": t["content"]} for t in tasks]


async def close_task(task_id: str) -> None:
    """Mark a Todoist task as complete (so it disappears from the list)."""
    async with httpx.AsyncClient(timeout=15.0) as client:
        r = await client.post(f"{API_BASE}/tasks/{task_id}/close", headers=_headers())
        r.raise_for_status()
